opt4j-core/src/main/java/org/opt4j/core/common/archive/AbstractArchive.java

```diff
@@ -54,17 +49,31 @@
 	@Override
 	public boolean update(Set<? extends Individual> individuals) {
 		List<Individual> candidates = new ArrayList<Individual>(individuals);
-		Objectives o1, o2;
-
 		/*
-		 * Remove all individuals that are already in the archive and those that
-		 * are dominated among the candidates
+		 * Remove all individuals that are already in the archive.
 		 */
 		candidates.removeAll(this);
+		removeDominatedCandidates(candidates);
+
+		removeArchiveDominated(candidates);
+
+		return updateWithNondominated(candidates);
+	}
+
+	/**
+	 * Remove candidates, which are weakly dominated by another candidate.
+	 * 
+	 * Thus, the list of candidates which need to be tested against the whole
+	 * archive is reduced.
+	 * 
+	 * @param candidates
+	 *            the list of candidates to sanitize
+	 */
+	void removeDominatedCandidates(List<Individual> candidates) {
 		for (int i = 0; i < candidates.size() - 1; i++) {
-			o1 = candidates.get(i).getObjectives();
+			Objectives o1 = candidates.get(i).getObjectives();
 			for (int j = i + 1; j < candidates.size(); j++) {
```

Comment:
How are you not running into an OutOfBoundsException here (list.get(list.size()))?